ambry-tools/src/main/java/com.github.ambry.tools/admin/ConcurrencyTestTool.java

```diff
@@ -0,0 +1,946 @@
+   */
+  class PutThread implements Runnable {
+    private final String threadName;
+    private final PutGetHelper producerJob;
+    private final ArrayList<BlobAccessInfo> blobInfos;
+    private final int maxPutCount;
+    private final AtomicInteger currentPutCount;
+    private final int maxParallelRequest;
+    private final long sleepTimeBetweenBatchPutsInMs;
+    private final boolean enableVerboseLogging;
+    private final ThreadLocalRandom threadLocalRandom;
+
+    public PutThread(String threadName, PutGetHelper producerJob, ArrayList<BlobAccessInfo> blobInfos,
+        final int maxPutCount, AtomicInteger currentPutCount, int maxParallelRequest,
+        long sleepTimeBetweenBatchPutsInMs, boolean enableVerboseLogging, ThreadLocalRandom threadLocalRandom) {
+      this.threadName = threadName;
+      this.producerJob = producerJob;
+      this.blobInfos = blobInfos;
+      this.maxPutCount = maxPutCount;
+      this.currentPutCount = currentPutCount;
+      this.maxParallelRequest = maxParallelRequest;
+      this.sleepTimeBetweenBatchPutsInMs = sleepTimeBetweenBatchPutsInMs;
+      this.enableVerboseLogging = enableVerboseLogging;
+      this.threadLocalRandom = threadLocalRandom;
+    }
+
+    @Override
+    public void run() {
+      while (currentPutCount.get() < maxPutCount) {
+        try {
+          int burstCount = threadLocalRandom.nextInt(maxParallelRequest) + 1;
+          logger.trace(threadName + " Producing " + burstCount + " times ");
+          CountDownLatch countDownLatch = new CountDownLatch(burstCount);
+          for (int j = 0; j < burstCount; j++) {
+            Callback callback = new Callback() {
+              @Override
+              public void onCompletion(Object result, Exception exception) {
+                if (exception == null) {
+                  Pair<String, byte[]> response = (Pair<String, byte[]>) result;
+                  if (currentPutCount.get() < maxPutCount) {
+                    blobInfos.add(new BlobAccessInfo(response.getFirst(), response.getSecond()));
+                    currentPutCount.incrementAndGet();
+                  }
+                }
+                countDownLatch.countDown();
+              }
+            };
+            FutureResult futureResult = new FutureResult();
+            producerJob.produce(futureResult, callback);
+          }
+          countDownLatch.await();
+          Thread.sleep(threadLocalRandom.nextLong(sleepTimeBetweenBatchPutsInMs));
+        } catch (InterruptedException ex) {
+          logger.error("InterruptedException in " + threadName + " " + ex.getStackTrace());
+        }
+      }
+      logger.info("Exiting Producer " + threadName + ": current Put Count " + currentPutCount.get() + ", max Put Count "
+          + maxPutCount);
+    }
+  }
+
+  /**
+   * Record to hold information about a blob along with its content and its reference count
+   */
+  class BlobAccessInfo {
+    String blobId;
+    AtomicInteger referenceCount;
```

Comment:
Each blob needs to be fetched "maxGetCountPerBlob" times before being deleted. so, this referenceCount tracks the same.